sm2/src/pke/decrypting.rs

```diff
@@ -0,0 +1,216 @@
+        self.decrypt_der_digest::<Sm3>(ciphertext)
+    }
+
+    /// Decrypts a ciphertext in-place from ASN.1 format using the specified digest algorithm.
+    pub fn decrypt_der_digest<D>(&self, ciphertext: &[u8]) -> Result<Vec<u8>>
+    where
+        D: 'static + Digest + DynDigest + Send + Sync,
+    {
+        let cipher = Cipher::from_der(ciphertext).map_err(elliptic_curve::pkcs8::Error::from)?;
+        let prefix: &[u8] = &[0x04];
+        let x: [u8; 32] = cipher.x.to_be_bytes();
+        let y: [u8; 32] = cipher.y.to_be_bytes();
+        let cipher = match self.mode {
+            Mode::C1C2C3 => [prefix, &x, &y, cipher.cipher, cipher.digest].concat(),
+            Mode::C1C3C2 => [prefix, &x, &y, cipher.digest, cipher.cipher].concat(),
+        };
+
+        Ok(self.decrypt_digest::<D>(&cipher)?.to_vec())
+    }
+}
+
+//
+// Other trait impls
+//
+
+impl AsRef<EncryptingKey> for DecryptingKey {
+    fn as_ref(&self) -> &EncryptingKey {
+        &self.encryting_key
+    }
+}
+
+impl ConstantTimeEq for DecryptingKey {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.secret_scalar.ct_eq(&other.secret_scalar)
+    }
+}
+
+impl Debug for DecryptingKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DecryptingKey")
+            .field("private_key", &self.secret_scalar.as_ref())
+            .field("encrypting_key", &self.encrypting_key())
+            .finish_non_exhaustive()
+    }
+}
+
+/// Constant-time comparison
+impl Eq for DecryptingKey {}
+impl PartialEq for DecryptingKey {
+    fn eq(&self, other: &DecryptingKey) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+fn decrypt(
+    secret_scalar: &Scalar,
+    mode: Mode,
+    hasher: &mut dyn DynDigest,
+    cipher: &[u8],
+) -> Result<Vec<u8>> {
+    let q = U256::from_be_hex(FieldElement::MODULUS);
+    let c1_len = (q.bits() + 7) / 8 * 2 + 1;
+
+    // B1: get 𝐶1 from 𝐶
+    let (c1, c) = cipher.split_at(c1_len as usize);
+    let encoded_c1 = EncodedPoint::from_bytes(c1).map_err(Error::from)?;
+
+    // verify that point c1 satisfies the elliptic curve
+    let mut c1_point = AffinePoint::from_encoded_point(&encoded_c1).unwrap();
+
+    // B2: compute point 𝑆 = [ℎ]𝐶1
+    let s = c1_point * Scalar::reduce(U256::from_u32(FieldElement::S));
+    if s.is_identity().into() {
+        return Err(Error);
+    }
+
+    // B3: compute [𝑑𝐵]𝐶1 = (𝑥2, 𝑦2)
+    c1_point = (c1_point * secret_scalar).to_affine();
+    let digest_size = hasher.output_size();
+    let (c2, c3) = match mode {
+        Mode::C1C3C2 => {
+            let (c3, c2) = c.split_at(digest_size);
+            (c2, c3)
+        }
+        Mode::C1C2C3 => c.split_at(c.len() - digest_size),
+    };
+
+    // B4: compute 𝑡 = 𝐾𝐷𝐹(𝑥2 ∥ 𝑦2, 𝑘𝑙𝑒𝑛)
+    // B5: get 𝐶2 from 𝐶 and compute 𝑀′ = 𝐶2 ⊕ t
+    let mut c2 = c2.to_owned();
+    kdf(hasher, c1_point, &mut c2)?;
+
+    // compute 𝑢 = 𝐻𝑎𝑠ℎ(𝑥2 ∥ 𝑀′∥ 𝑦2).
+    let mut u = vec![0u8; digest_size];
```

Comment:
Similar comment to [this one](https://github.com/RustCrypto/elliptic-curves/pull/1069/files#r1733348316) regarding the dynamic allocations. They're unnecessary.